src/Components/Components/src/PersistentState/PersistentValueProviderComponentSubscription.cs

```diff
@@ -233,10 +233,28 @@ private static PropertyGetter PropertyGetterFactory((Type type, string propertyN
         {
             throw new InvalidOperationException($"Property {propertyName} not found on type {type.FullName}");
         }
+
+        // Check if the property is public
+        if (propertyInfo.GetMethod == null || !propertyInfo.GetMethod.IsPublic)
+        {
+            throw new InvalidOperationException(
+                $"The property '{propertyName}' on component type '{type.FullName}' cannot be used with PersistentState because it is not public. Properties with PersistentState must be public.");
+        }
+
         return new PropertyGetter(type, propertyInfo);
 
         static PropertyInfo? GetPropertyInfo([DynamicallyAccessedMembers(LinkerFlags.Component)] Type type, string propertyName)
-            => type.GetProperty(propertyName);
+        {
+            // First try to find the property with public access only (for performance in the common case)
+            var publicProperty = type.GetProperty(propertyName);
+            if (publicProperty != null)
+            {
+                return publicProperty;
+            }
+
+            // If not found as public, try with all flags to see if it exists as private/protected
+            return type.GetProperty(propertyName, ComponentProperties.BindablePropertyFlags);
+        }
```

Comment:
Undo this change if it's not needed (if the property is not public it will return null)